sdks/java/io/elasticsearch-tests/elasticsearch-tests-common/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTestCommon.java

```diff
@@ -512,4 +527,44 @@ public Void apply(Iterable<String> input) {
       return null;
     }
   }
+
+  /** Test that the default predicate correctly parses chosen error code. */
+  public void testDefaultRetryPredicate(RestClient restClient) throws IOException {
+    assertFalse(DEFAULT_RETRY_PREDICATE.test(new IOException("test")));
+    String x =
+        "{ \"index\" : { \"_index\" : \"test\", \"_type\" : \"doc\", \"_id\" : \"1\" } }\n"
+            + "{ \"field1\" : @ }\n";
+    HttpEntity entity = new NStringEntity(x, ContentType.APPLICATION_JSON);
+
+    Response response = restClient.performRequest("POST", "/_bulk", Collections.emptyMap(), entity);
+    assertTrue(CUSTOM_RETRY_PREDICATE.test(new ResponseException(response)));
+  }
+
+  /**
+   * Test that retries are invoked when Elasticsearch returns a specific error code. We invoke this
+   * by issuing corrupt data and retrying on the `400` error code. Normal behaviour is to retry on
+   * `429` only but that is difficult to simulate reliably. The logger is used to verify expected
```

Comment:
@echauchot the test is interesting, but i am afraid i can't think of a way to simulate that condition(since the retry code is internal and i cant change the request when the pipeline is executing). Please suggest !

Reply:
Indeed very difficult to simulate with an http 400 bad formatted doc. The batch will be retried until maxAttempts is reached because it will still contain a bad formatted doc. The only way would be to test with an error that can be temporary (like an ES status) and control the embeded ES through a separate thread in the UT to change that status. But now that you changed the logic in `handleRetry()` and considering that you test your predicate in a separate UT, I think it is fine.